src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -771,23 +795,134 @@ private RegexNode ReduceConcatenation()
             {
                 children.RemoveRange(j, i - j);
             }
+        }
 
-            // Now try to convert as many loops as possible to be atomic to avoid unnecessary backtracking.
-            if ((Options & RegexOptions.RightToLeft) == 0)
+        /// <summary>
+        /// Combine adjacent loops.
+        /// a*a*a* => a*
+        /// </summary>
+        private void ReduceConcatenationWithAdjacentLoops()
+        {
+            Debug.Assert(Type == Concatenate);
+            Debug.Assert(Children is List<RegexNode>);
+
+            var children = (List<RegexNode>)Children!;
+            int current = 0, next = 1, nextSave = 1;
+
+            while (next < children.Count)
             {
-                ReduceConcatenateWithAutoAtomic();
+                RegexNode currentNode = children[current];
+                RegexNode nextNode = children[next];
+
+                if (currentNode.Options == nextNode.Options)
+                {
+                    static bool CanCombineCounts(int nodeMin, int nodeMax, int nextMin, int nextMax)
+                    {
+                        // We shouldn't have an infinite minimum; bail if we find one. Also check for the
```

Comment:
“Shouldn’t “ meaning it would be a bug in us? If so, you’ve used Debug.Assert for those elsewhere.

Reply:
I didn't use an assert because I'm not 100% certain it's not possible. I'll investigate more subsequently.